src/main/java/com/google/sps/data/TimeRange.java

```diff
@@ -1,4 +1,195 @@
+  }
+
+  /**
+   * Checks if this range completely contains another range. This means that {@code other} is a
+   * subset of this range. This is an inclusive bounds, meaning that if two ranges are the same,
+   * they contain each other.
+   */
+  public boolean contains(TimeRange other) {
+    /** If this range has no duration, it is irrelevant. */
+    if (Duration.between(start, end).isZero()) {
+      return false;
+    }
+
+    /** If the start of one range is after the end of another, the range isn't contained. */
+    if (other.start.isAfter(this.end)) {
+      return false;
+    }
+
+    /** If the the end of one range is before the start of another, the range isn't contained. */
+    if (other.end.isBefore(this.start)) {
+      return false;
+    }
+
+    /** If the end of one range is equal to the start of another, check whether the start of one range comes after the other. */
+    if (other.end == this.start) {
+      return other.start.isAfter(this.start);
+    }
+
+    /** If the ends of the range are equal, check whether the start of range comes after the other. */
+    if (other.end == this.end) {
+      return other.start.isAfter(this.start);
+    }
+
+    /** If the end of one range is after the end of another the range is not contained. */
+    if (other.end.isAfter(this.end)) {
+      return false;
+    }
+
+    /** If the start of the ranges are the same, check if one end is before the other. */
+    if (other.start == this.start) {
+      return other.end.isBefore(this.end);
+    }
+
+    /**
+     * If the other range has no duration, then we must treat it like a point in time rather than a
+     * range.
+     */
+    if (Duration.between(other.start, other.end).isZero()) {
+      return contains(this, other.start);
+    }
+
+    // We need the inclusive end for this check in order for this case to equal true:
+    // |------|
+    //     |--|
+    Instant otherInclusiveEnd = other.end;
+    return contains(this, other.start) && contains(this, otherInclusiveEnd);
+  }
+
+  /** Checks if a timerange contains a certain instant. */
+  public boolean contains(Instant point) {
+    return contains(this, point);
+  }
+
+  @Override
+  public boolean equals(Object other) {
+    return other instanceof TimeRange && equals(this, (TimeRange) other);
+  }
+
+  private static boolean contains(TimeRange range, Instant instant) {
+    /** If a range has no duration, it cannot contain anything. */
+    if (Duration.between(range.start, range.end).isZero()) {
+      return false;
+    }
+
+    /** If the point comes before the start of the range, the range might not contain it. */
+    if (instant.isBefore(range.start)) {
+      return false;
+    }
+
+    /** If the point comes after the end of the range, the range might not contain it. */
+    if (instant.isAfter(range.end)) {
+      return false;
+    }
+
+    /** If the point is equal to the end of the range, the range might not contain it. */
+    if (instant == range.end) {
```

Comment:
This `instant == range.end` undermines what you want here:
`// We need the inclusive end for this check in order for this case to equal true:
    // |------|
    //     |--|` (lines 141 ish)

For simpler code, l'd suggest you think about how your `contains(TimeRange range, Instant instant)` could be simplified with this logic: 
`
!instant.isBefore(range.start) && !instant.end.isAfter(range.end)
`
Then instant must be >= range.start and instant's end must be <= range.end

Reply:
Instant doesn't have an end it represents a start or an end depending on the what is passed to the method

Reply:
In this case the instant could represent the start of a range and if the start is equal to the end of another range, they range is not contained

Reply:
Sorry l meant `!instant.isBefore(range.start) && !instant.isAfter(range.end)`
The logic is right here. The `contains(TimeRange other)` is where things get tricky. See my comment there.